add initial implementation for mounting Azurefile on windows node

andyzhangx opened the pull request:
**What this PR does / why we need it**:
add initial implementation for mounting Azurefile on windows node.


**Special notes for your reviewer**:
As there is a known blocking issue in Windows 2016 server when mounting a SMB share on a container host and then bind that share to a container ( Azure file on Windows is using SMB protocol), this PR still could not mount an azure file on current windows 2016 server node, it depends on 2016 RS3 release, and it will still succeed (as a workaround) if customer want to mount an azure file on current windows node.

Main code logic is similar to what it does in Linux node:
1) create target directory  in Windows host
2) Use New-SmbGlobalMapping powershell cmdlet to mount SMB azure file to a drive in Windows host
3) Use mklink command to link target directory to the mounted driver
4) K8s would bind target directory to the container directory

source in mount function would be like:
\\\\[accountname].file.core.windows.net\test

target in mount function would be like:
c:\var\lib\kubelet\pods\5f679f75-7ce3-11e7-b718-000d3a31dac4\volumes\kubernetes.io~azure-file

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note

andyzhangx committed:
initial work on enabling azure file on windows

andyzhangx committed:
add initial work for mount azure file on windows

dixudx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+	parentDir := getWindowsParentDir(target)
+	err := os.MkdirAll(parentDir, 0700)
+	if err != nil {
+		glog.Infof("mkdir(%s) failed: %v\n", parentDir, err)
```
`s/(%s)/%q`

andyzhangx replied:
done

dixudx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+	} else {
+		glog.Infof("Unmount succeeded, output: %s", output)
+	}
+	return nil
```
`return err`

andyzhangx replied:
fixed

dixudx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func getWindowsParentDir(path string) string {
+	index := strings.LastIndex(path, "\\")
```
use `os.PathSeparator` instead of this hard-coded way

dixudx replied:
Why not import `path` and use `path.Dir(pathName)`.

andyzhangx replied:
I have used os.PathSeparator, path.Dir(pathName) only works in linux, so I must implement this func in the code here

dixudx replied:
`os.PathSeparator` works on [Windows](https://golang.org/src/os/path_windows.go).

Also [`path.Dir`](https://golang.org/src/path/filepath/path.go#L457).

andyzhangx replied:
@dixudx thanks for your point, actually filepath.Dir works for windows, path.Dir only works in Linux because in the end it will use below func:
func Split(path string) (dir, file string) {
	i := strings.LastIndex(path, "/")
	return path[:i+1], path[i+1:]
}

andyzhangx replied:
so I will remove getWindowsParentDir, it's not necessary now, thanks.

dixudx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (mounter *Mounter) IsMountPointMatch(mp MountPoint, dir string) bool {
+	return (mp.Path == dir)
```
redundant parenthesis

andyzhangx replied:
It's more like a default implementation here, may change later

dixudx commented on pkg/volume/azure_file/azure_file.go:
```diff
@@ -208,15 +209,24 @@ func (b *azureFileMounter) SetUpAt(dir string, fsGroup *int64) error {
+	source := ""
+
+	if runtime.GOOS == "windows" {
+		source = fmt.Sprintf("\\\\%s.file.%s\\%s", accountName, getStorageEndpointSuffix(b.plugin.host.GetCloudProvider()), b.shareName)
```
use `os.PathSeparator`

andyzhangx replied:
fixed

dixudx commented on pkg/volume/azure_file/azure_file.go:
```diff
@@ -208,15 +209,24 @@ func (b *azureFileMounter) SetUpAt(dir string, fsGroup *int64) error {
+		source = fmt.Sprintf("\\\\%s.file.%s\\%s", accountName, getStorageEndpointSuffix(b.plugin.host.GetCloudProvider()), b.shareName)
+		mountOptions = []string{"use", "*", source, fmt.Sprintf("/u:AZURE\\%s", accountName), accountKey, "/p:yes"}
+	} else {
+		os.MkdirAll(dir, 0700)
```
0755

andyzhangx replied:
Hi, I did not change the original value 0700, is there any reason to change it to 0755. Seems 0700 is more safe as it's only accessible by root user.

dixudx commented on pkg/volume/azure_file/azure_file.go:
```diff
@@ -208,15 +209,24 @@ func (b *azureFileMounter) SetUpAt(dir string, fsGroup *int64) error {
+		os.MkdirAll(dir, 0700)
+		source = fmt.Sprintf("//%s.file.%s/%s", accountName, getStorageEndpointSuffix(b.plugin.host.GetCloudProvider()), b.shareName)
+		// parameters suggested by https://azure.microsoft.com/en-us/documentation/articles/storage-how-to-use-files-linux/
+		options := []string{fmt.Sprintf("vers=3.0,username=%s,password=%s,dir_mode=0700,file_mode=0700", accountName, accountKey)}
```
better using `dir_mode=0755,file_mode=0644`?

andyzhangx replied:
let's keep the original mode

jdumars commented:
/sig windows
/sig azure

andyzhangx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+
+func (mounter *Mounter) Mount(source string, target string, fstype string, options []string) error {
+	if source == "tmpfs" || len(target) < 3 {
+		glog.Infof("Skip mounting source (%s), target (%s)\n, with arguments (%s) in windows", source, target, options)
```
done

andyzhangx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+		target = "c:" + target
+	}
+	parentDir := getWindowsParentDir(target)
+	err := os.MkdirAll(parentDir, 0700)
```
done

andyzhangx commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,122 @@
+
+	_, err = exec.Command("cmd", "/c", "mklink", "/D", target, source).CombinedOutput()
+	if err != nil {
+		glog.Infof("mklink failed: %v\n", err)
```
fixed

andyzhangx committed:
fix review comments

andyzhangx committed:
full implementation for attach azure file on windows node

andyzhangx committed:
working azure file mount

andyzhangx committed:
remove useless functions

andyzhangx commented:
@brendanburns @jdumars 
I have completed initial framework for mounting Azurefile on windows node
A more complete implementation code is now under:
andyzhangx/kubernetes@15c0975
While I could not let above code in because 2016 RS3 is not ready on azure, it will block existing logic when creating a pod with azure file mount on windows node. I will send a new PR when 2016 RS3 is ready on azure.

andyzhangx committed:
add a workable implementation about mounting azure file on windows node

andyzhangx commented:
@karataliu for code review

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+	parentDir := filepath.Dir(target)
+	err := os.MkdirAll(parentDir, 0755)
+	if err != nil {
+		glog.Infof("mkdir(%q) failed: %v\n", parentDir, err)
```
glog.Errorf

andyzhangx replied:
I changed to following, it's more reasonable:
return fmt.Errorf("mkdir(%q) failed: %v", parentDir, err)

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	if len(options) != 1 {
+		glog.Warningf("azureMount: mount options(%q) command(%n) not equal to 1, skip mounting.\n", options, len(options))
```
'\n' not needed. Also for other glog.__f

`Warningf logs to the WARNING and INFO logs. Arguments are handled in the manner of fmt.Printf; a newline is appended if missing.`

andyzhangx replied:
fixed

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+	}
+	cmd := options[0]
+
+	driverLetter, err := getAvailabeDriverLetter()
```
typo: Available

andyzhangx replied:
fixed

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+
+func getAvailabeDriverLetter() (string, error) {
+	cmd := "$used = Get-PSDrive | Select-Object -Expand Name | Where-Object { $_.Length -eq 1 }"
+	cmd += ";$drive = 67..90 | ForEach-Object { [string][char]$_ } | Where-Object { $used -notcontains $_ } | Select-Object -First 1;$drive"
```
65..90? A:/B: should be usable.

andyzhangx replied:
I would suggest now use these two drive letters

karataliu commented on pkg/volume/azure_file/azure_file.go:
```diff
@@ -208,15 +209,27 @@ func (b *azureFileMounter) SetUpAt(dir string, fsGroup *int64) error {
+	if runtime.GOOS == "windows" {
+		cmd := fmt.Sprintf(`$User = \"AZURE\\%s\";$PWord = ConvertTo-SecureString -String \"%s\" -AsPlainText -Force;
+			$Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $User, $PWord`,
+			accountName, accountKey)
```
I'd suggest passing 'username=%s,password=%s' only, and leave implementation to mounter.

Those directives have enforced underlying implementation, and it's for SMB protocol only.

The underlying mounter implementation is based on OS, so for windows, there's only one implementation (to mount AzureFile/AzureDisk or other kinds).

The problem is windows does not have 'mount' option semantic. We should be careful here to introduce new ones.

A potential reference is mount cmd for NFS on Windows:
https://technet.microsoft.com/en-us/library/cc754350(v=ws.11).aspx

andyzhangx replied:
good point, I will change it, only pass username and password

andyzhangx replied:
fixed

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	if !strings.HasPrefix(target, "c:") && !strings.HasPrefix(target, "C:") {
+		target = "c:" + target
```
User may be passing '/' in path. If we expect 'C:\a\b\' ahead, shall we do a normalization for target here.

https://forums.docker.com/t/volume-mounts-in-windows-does-not-work/10693/6

andyzhangx replied:
It's already normaliazed in k8s common code, like below:
("\\var\\lib\\kubelet\\pods\\f6590c44-7e6b-11e7-8f44-000d3a31dac4\\volumes\\kubernetes.io~secret\\default-token-rrrc5") 
while it's missing drive letter in the beginning.

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+	driverPath := driverLetter + ":"
+	cmd += fmt.Sprintf(";New-SmbGlobalMapping -LocalPath %s -RemotePath %s -Credential $Credential", driverPath, source)
+
+	_, err = exec.Command("powershell", "/c", cmd).CombinedOutput()
```
I'd suggest have a small func 'powershellRun' and also port 'cmd' to powershell cmdlets, in order to be more unified.

andyzhangx replied:
there are only two powershell command invocation now, I will add if there is more

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+func (mounter *Mounter) PathIsDevice(pathname string) (bool, error) {
+	return true, nil
```
should be false, Windows does not have device path.

andyzhangx replied:
fixed

karataliu commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,131 @@
+	return nil
+}
+
+func (mounter *SafeFormatAndMount) diskLooksUnformatted(disk string) (bool, error) {
```
not used anywhere.

Used to be linux only, and has renamed, should remove.
https://github.com/kubernetes/kubernetes/pull/16948
https://github.com/kubernetes/kubernetes/pull/31515

andyzhangx replied:
removed

andyzhangx committed:
fix review comments and make the pod creating successful even azure file mount failed

karataliu commented on pkg/volume/azure_file/azure_file.go:
```diff
@@ -208,15 +209,27 @@ func (b *azureFileMounter) SetUpAt(dir string, fsGroup *int64) error {
-	if b.readOnly {
-		options = append(options, "ro")
+	mountOptions := []string{}
+	source := ""
```
combine with Line 216

andyzhangx replied:
combined

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	if source == "tmpfs" {
+		glog.Infof("azureMount: mounting source (%q), target (%q)\n, with options (%q)", source, target, options)
```
probably don't need 'azureMount' maybe just 'windowsMount'?

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+
+	parentDir := filepath.Dir(target)
+	err := os.MkdirAll(parentDir, 0755)
+	if err != nil {
```
inline:
```go
if err:= ...; err != nil {
   ...
}
```

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+
+	if source == "tmpfs" {
+		glog.Infof("azureMount: mounting source (%q), target (%q)\n, with options (%q)", source, target, options)
+		os.MkdirAll(target, 0755)
```
You should `return os.MkdirAll(...)` in case that returns an error.

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	parentDir := filepath.Dir(target)
+	err := os.MkdirAll(parentDir, 0755)
+	if err != nil {
+		return fmt.Errorf("mkdir(%q) failed: %v", parentDir, err)
```
Don't wrap errors, just `return err`

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	if len(options) != 1 {
+		glog.Warningf("azureMount: mount options(%q) command(%n) not equal to 1, skip mounting.", options, len(options))
```
same wrt `azureMount`

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+
+	if len(options) != 1 {
+		glog.Warningf("azureMount: mount options(%q) command(%n) not equal to 1, skip mounting.", options, len(options))
+		return nil
```
Shouldn't we return an error here?

andyzhangx replied:
yes, should return error

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	cmd += fmt.Sprintf(";New-SmbGlobalMapping -LocalPath %s -RemotePath %s -Credential $Credential", driverPath, source)
+
+	_, err = exec.Command("powershell", "/c", cmd).CombinedOutput()
+	if err != nil {
```
inline:

```go
if _, err := exec.Command(...); err != nil {
...
}
```

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+import (
+	"fmt"
+	"os"
+	"os/exec"
```
We have a wrapper on `os/exec` here:

https://github.com/kubernetes/utils/blob/master/exec/exec.go

Please use that, it makes testing this stuff _much_ easier...

andyzhangx replied:
fixed, I have used "k8s.io/utils/exec"

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+		// we don't return error here, even though New-SmbGlobalMapping failed, we still make it successful,
+		// will return error when Windows 2016 RS3 is ready on azure
+		glog.Errorf("azureMount: SmbGlobalMapping failed: %v", err)
+		os.MkdirAll(target, 0755)
```
as above wrt returning the error from `MkdirAll`

andyzhangx replied:
fixed

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+
+	_, err = exec.Command("cmd", "/c", "mklink", "/D", target, driverPath).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("mklink failed: %v", err)
```
don't wrap errors.

andyzhangx replied:
fixed

andyzhangx replied:
After my debugging, I found err return insufficient info, we must return output info to locate the issue, below are my wrapping way:
return fmt.Errorf("mklink failed: %v, output: %q", err, string(output))

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	_, err = exec.Command("cmd", "/c", "mklink", "/D", target, driverPath).CombinedOutput()
+	if err != nil {
```
inline:
```go
if _, err := exec.Command(...); err != nil {
   ...
}
```

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	glog.Infof("azureMount: Unmount target (%q)", target)
+	output, err := exec.Command("cmd", "/c", "rmdir", target).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Unmount failed: %v", err)
```
don't wrap errors.

andyzhangx replied:
fixed

andyzhangx replied:
After my debugging, I found err return insufficient info, we must return output info to locate the issue, below are my wrapping way:
return fmt.Errorf("rmdir failed: %v, output: %q", err, string(output))

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+func (mounter *Mounter) IsLikelyNotMountPoint(file string) (bool, error) {
+	return true, nil
```
Is there a way to implement this in Windows?

andyzhangx replied:
difficult point is in windows, it does not have (*syscall.Stat_t).Dev to locate the device, I will try to find similar implementation then

andyzhangx replied:
Finally I got a way to do that:
use filemode to check whether current file is a symlink, if yes, then it's not a mount point

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,132 @@
+	cmd := "$used = Get-PSDrive | Select-Object -Expand Name | Where-Object { $_.Length -eq 1 }"
+	cmd += ";$drive = 67..90 | ForEach-Object { [string][char]$_ } | Where-Object { $used -notcontains $_ } | Select-Object -First 1;$drive"
+	output, err := exec.Command("powershell", "/c", cmd).CombinedOutput()
+	if err != nil || len(output) == 0 {
```
Differentiate between these two cases:
```go
if err != nil {
   return err
}
if len(output) == 0 {
   return fmt.Errorf(...)
}
```

andyzhangx replied:
fixed

brendandburns commented:
This needs unit tests. Switching to the `util/exec` package will help make that easier.

andyzhangx committed:
fix according to review comments

andyzhangx committed:
add mount_windows_test

andyzhangx committed:
add implementation for IsLikelyNotMountPoint func

andyzhangx commented:
@brendanburns I have switched to use util/exec package and added a new unit test file(mount_windows_test.go), while by default, only linux unit tests would run.

jdumars commented:
/ok-to-test

andyzhangx commented:
/retest

andyzhangx commented:
/retest

andyzhangx committed:
remove mount_windows_test.go temporaly

andyzhangx committed:
add back unit test for mount_windows.go

andyzhangx committed:
add normalizeWindowsPath func

andyzhangx commented:
@brendanburns do you have time to review, thanks.
And I am now working on azure disk on windows feature, it depends on this PR merging first.
@jdumars @kyliel for awareness.

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	if output, err := ex.Command("cmd", "/c", "mklink", "/D", target, driverPath).CombinedOutput(); err != nil {
+		return fmt.Errorf("mklink failed: %v, output: %q", err, string(output))
```
please don't wrap errors...

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,151 @@
+	target = normalizeWindowsPath(target)
+	ex := exec.New()
+	if output, err := ex.Command("cmd", "/c", "rmdir", target).CombinedOutput(); err != nil {
+		return fmt.Errorf("rmdir failed: %v, output: %q", err, string(output))
```
Here too.

brendandburns commented on pkg/util/mount/mount_windows.go:
```diff
@@ -0,0 +1,151 @@
+	ex := exec.New()
+	output, err := ex.Command("powershell", "/c", cmd).CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("getAvailableDriveLetter failed: %v, output: %q", err, string(output))
```
here too (don't wrap errors)

brendandburns commented:
Some more comments about wrapping `error` objects. In general, I'm very negative on wrapping errors because you lose most of the structure (unlike in java).

So if you want detailed info `log.errorf` it but return the `err` back unchanged.

Other than that, lgtm.

andyzhangx commented:
Hi @brendanburns thanks for the review.
I have to wrap errors  in powershell invocation, because err does not give sufficient info, the detailed error info is all in output.

andyzhangx committed:
fix normalizeWindowsPath func issue

andyzhangx commented:
/retest

brendandburns commented:
/retest

andyzhangx commented:
this PR moved to https://github.com/kubernetes/kubernetes/pull/52469
